### DISCUSSION TITLE
How to report events in Stripe events handler

### DISCUSSION THREAD

**Discussion by koteisaev** [plausible/analytics#1845] [Anything Else]
How to report events from Stripe events handler?

The way how visitor counted at Plausible Analytics (request `IP`+`User-Agent` used as user id of sorts) gives interesting impact on scenarios like stripe events handler.

We can send conversion events from browser, but events like `purchase` about actual payments can be sent only from Stripe events handler, since we are subscription-based service with a free trial period of 10 days.

Does it mean we will have to save user `ip` and `user-agent` of customer at moment of conversion and send it with purchase event?
Or there is other approach we can use in such context to preserve attribution of purchases to location and device type and os/browser brand and version? Can we override these location and device information for the event?

What is best way to attribute purchases acts (actual payments events) from from Stripe events handler to standard UTM tags?
As I see this is done via some virtual pageview like `https://example.com/billing/events/purchase/stripe` with some metadata?

What is best way to use a custom attribution across events (pageviews, conversions, purchases) with our url parameters like `link_id` and `partner_id`  both in browser and in stripe handler?

**Comment by RobertJoonas** [comment]
Hi @koteisaev!

I think the solution here would be getting the conversion data back from Stripe and sending the `payment` event to Plausible after the user has been redirected back to your site. The easiest setup is using [pageview goals](https://plausible.io/docs/pageview-goals). For that you should have a certain page that the user lands on after the payment (say, a thank-you page).

Then you would know what browser, OS, screen size, ... (+ all other session metrics) the visitor used, who visited this thank-you page. This is actually how we have our own subscription tracking set up. You can see it here on our live demo: https://plausible.io/plausible.io?goal=Visit+%2Fbilling%2Fupgrade-success. 

If you also want to send some custom properties (say amount for example), then you should use [custom events](https://plausible.io/docs/custom-event-goals). It should also be possible to get the payment data back from Stripe once the purchase has been completed. Then you could send the additional properties as custom event props.

Happy to advise should you have any more questions. Feel free to also reach out via email so that we can have a closer look at your setup. Thanks!

**Reply by koteisaev** [reply]
@RobertJoonas I provided a detailed explanation of some Stripe related aspects in to discussion reply, but want again summarize it here in directly reply to your answer, and also add some missing information.

* We use a wizard like page, without redirects, with embedded Stripe Elements, and to this regard sending conversion or upgrade event will not be an issue.
* But we use subscriptions so actual payments happen outside of our website in few days after conversion, and there is single place we can send events about payments to attribute them to UTM tags and a custom parameter like `partner_id` is Stripe events handler where normal browser context is not available.
* Stripe provides customer last known IP and pre-parsed browser and OS name & version information, so raw `user agent` value does not available.

**Comment by koteisaev** [comment]
Hello!
Thanks for your reply!

I reviewed what Stripe provides among details about payment intent in related events, e. g. successful payment in context of payment intent.

It provides ready (already parsed) user agent information like:

```json
{
"browser": "Chrome",
"data_source": "client",
"device": "PC",
"platform": "Windows 10",
"session_id": "SOME_SESSION_UNIQUE_ID",
"show_session_data": true,
"version": "100.0.4896.127"
}
```

Also Stripe Payment intent contains (probably last known by Stripe) customer IP address and resolved geolocation data.
like this:

```json
{
"city_name": "K'alak'i T'bilisi"
"country": "GE"
"country_code": "GE"
"country_name": "Georgia"
"lat": 41.71629999999999
"lon": 44.83869999999999
"region_name": "51"
"value": "SOME_IP_VALUE"
}
```
The `country`. `country_name`,  `region_name`, and `city_name` seems to be their DB internal values at stripe, Especially `region_name`.

So, if I could somehow send event with `user-agent` like 
```
example.com (Nodejs/14.0; Stripe-Client/9.0; Implementation/1.0.3;Use-Override/1)
```

and use some custom parameters at events like

```json
{
"$override_ip":"SOME_IP_VALUE",
"$override_browser_name": "Chrome",
"$pverride_browser_version": "100.0.4896.127"
"$override_os_name":"Windows",
"$override_os_version":10
"$override_device_type":"PC",
}
```

it may help me to solve my problem with attribution of events coming from Stripe events handler..

So far, I do not have other option except saving user agent and user IP at moment of conversion/subscription/upgrade and use this info for events related to specific subscription to some extent. If that information is not available, e. g. event related to some old subscription, I can not send this request to Plausible analytics because I do not know what to be used e. g. as stub `user-agent` and `IP` .

**Reply by ukutaht** [reply]
> But we use subscriptions so actual payments happen outside of our website in few days after conversion, 

Honestly this seems like more of a 'Product Analytics' use-case than 'Web Analytics'. The big difference is that in Product Analytics tools like PostHog for example, you can provide a user id from your own database. That way you can get long-term retention information and correlating things like trial signup with a payment multiple days or weeks later.

Plausible is designed to time out a session after 30 minutes, and the generated user IDs change every 24 hours even if you provide the same IP and User-Agent. It's suitable for more ephemeral data like anonymous (= non-logged-in) visitors. Plausible is not capable or correlating a payment event to a visit that happened days previously.

**Reply by koteisaev** [reply]
What matters for me now is ability to bind things like country, browser/os, coupon, UTM tags, our private tags like partner id and link id, etc. to purchases, and group purchases by them. In that sense, I do not need exact user id here. For cases where user id matters, I agree that for this kind of tasks, like funnel visualizations, I will need something else - from own-made solution to some other product like PostHog...
I see single problem with such approach about using Plausible to store attributed aggregated purchase events data is possible distortion (some increase) of unique visitors and overall visits) in site stats. But hope this distortion is not so strong because (sadly, sic!) purchases/payments happen much less often than page visits.
